### DISCUSSION TITLE
Optimal way to write sparse console output

### DISCUSSION THREAD

**Discussion by alabuzhev** [microsoft/terminal#16208] [Q&A]
TL;DR: an app needs to update various potions of the screen from time to time, typically due to some user actions. To avoid flickering and improve performance, the app uses double buffering, i.e. all writes go to a memory buffer, which is then flushed to the console periodically. The changed areas are not necessarily contiguous or adjacent, so either multiple host calls must be made or extra areas have to be updated, up to and including the whole screen. The question is about finding the perfect balance between the number of host calls and the number of screen cells to update.

Let's say we have this nice default state initially:
![1](https://github.com/microsoft/terminal/assets/11453922/4a79ca91-c883-4a7c-be93-a3d9a0d4cfe1)

And need to draw this (red areas represent some generic colored text the user wants to see):
![2](https://github.com/microsoft/terminal/assets/11453922/b2bf85ca-848d-43a2-9914-84c05457fc6b)

We can traverse the buffer line by line, find modified blocks and send each of them to the host via WriteConsoleOutput or WriteConsole with some VT seasoning:
![3_13_calls](https://github.com/microsoft/terminal/assets/11453922/32e5df58-369b-48af-88c5-15f114eda8c5)

The number of such calls, however, will be quite high, 13 in this case and likely up to a hundred in real world scenarios.
We can optimize it by merging the adjacent rectangles:
![4_10_calls](https://github.com/microsoft/terminal/assets/11453922/d356ed5a-353a-4c3b-b017-12fa9d5d25e8)

The number of calls is still relatively high, because perfectly aligned blocks are rare. To decrease it further, we can increase tolerance and include some unmodified cells to merge more, resulting in 5 larger rectangles and 5 host calls:
![5_5_calls](https://github.com/microsoft/terminal/assets/11453922/05049e4f-3038-4c93-a285-95546e562fc7)

Going further, we can say "to hell with it, if there are any changes we're updating the whole line", and get something like:
![6_2_calls](https://github.com/microsoft/terminal/assets/11453922/a6b42a89-f827-4127-8cbf-b79135e3c4f2)

And, ultimately, we can redraw the whole screen in one call if there are any changes anywhere without all this nonsense:
![7_1_call](https://github.com/microsoft/terminal/assets/11453922/9d02815b-8956-46f0-8334-c1c8a3390a57)

The question is, which way is ultimately cheaper and more preferable - "more calls, less data" or "fewer calls, more data"?

I tried to measure it of course, but the differences are not that pronounced to make a conclusion (except that redrawing the whole screen apparently isn't as bad as one might expect).

Perhaps you have some internal stats, metrics, observations or general guidelines?

P.S. I suppose these days all the initial blocks can be sent in just one WriteConsole call with multiple VT cursor positioning instructions, but for older OSes we still need to support WriteConsoleOutput path that only works with individual rectangles.

**Comment by lhecker** [comment]
`big.txt` has 6'488'666 characters. Printing it all at once with `WriteFile` on my Windows 11 machine (build 25983), conhost takes 0.18s (35MB/s). Printing it character-by-character takes ~90s (72kB/s). It should be in the same order of magnitude all the way down to Windows XP, even on worse HW. In any case, this means that the overhead per call is somewhere around 90s/6488666 = 14us and that the overhead per character is around 0.18s/6488666 = 0.028us. I haven't tested `WriteConsoleOutput` but I suspect that it performs about the same.

In other words, if you know you need to print at least once (~14us), then a second, separate print (~28us total) is only cheaper if it avoids rewriting at least 14us/0.028us = 500 unrelated characters. I don't think dirtying too many cells is a concern, because GDI honestly isn't _that_ slow and realistically most people have a GPU.

If VT support is available however, I'd always use CUP sequences, because they're comparatively very cheap and can draw disjoint regions in a single call.
